src/logplex_message.erl

```diff
@@ -5,40 +5,109 @@
+-export([process_msgs/4
          ,process_msgs/1
          ,process_error/5
          ,process_error/4
-         ,parse_msg/1
         ]).
 
 -include("logplex.hrl").
 -include("logplex_error.hrl").
 -include("logplex_logging.hrl").
--define(SI_KEY, logplex_redis_buffer_map).
 
-shard_info() ->
-    logplex_shard_info:read(?SI_KEY).
+%% ----------------------------------------------------------------------------
+%% API Functions
+%% ----------------------------------------------------------------------------
 
 process_msgs(Msgs, ChannelId, Token, TokenName) when is_list(Msgs) ->
+    case logplex_app:config(batch_redis, false) of
+        true ->
+            process_msgs_batch_redis(Msgs, ChannelId, Token, TokenName);
+        _ ->
+            process_msgs_classic(Msgs, ChannelId, Token, TokenName)
+    end.
+
+process_msgs_classic(Msgs, ChannelId, Token, TokenName) ->
     ShardInfo = shard_info(),
     [ process_msg(RawMsg, ChannelId, Token, TokenName, ShardInfo)
       || RawMsg <- Msgs ],
     ok.
 
+process_msgs_batch_redis(Msgs, ChannelId, Token, TokenName) when is_list(Msgs) ->
+    RawMsgs = get_raw_msgs(Msgs),
+    logplex_stats:incr(message_received, length(RawMsgs)),
+    logplex_realtime:incr('message.received', length(RawMsgs)),
+    case logplex_channel:lookup_flag(no_redis, ChannelId) of
+        not_found ->
+            logplex_realtime:incr(unknown_channel, length(RawMsgs)),
+            ?INFO("at=process_msgs channel_id=~s msg=unknown_channel", [ChannelId]),
+            ok;
+        Flag ->
+            [logplex_firehose:post_msg(ChannelId, TokenName, RawMsg) || RawMsg <- RawMsgs],
+            CookedMsgs = [ iolist_to_binary(re:replace(RawMsg, Token, TokenName)) || RawMsg <- RawMsgs],
+
+            [process_drains(ChannelId, CookedMsg, logplex_app:config(deny_drain_forwarding, false)) || CookedMsg <- CookedMsgs],
+            [process_tails(ChannelId, CookedMsg, logplex_app:config(deny_tail_sessions, false)) || CookedMsg <- CookedMsgs],
+
+            process_redis_batch(ChannelId, CookedMsgs, logplex_app:config(deny_redis_buffers, Flag)),
+            ok
+    end.
+
 process_msgs(Msgs) ->
-    ShardInfo = shard_info(),
-    [ process_msg(RawMsg, ShardInfo) || RawMsg <- Msgs ].
+    case logplex_app:config(batch_redis, false) of
+        true ->
+            [begin
+                 logplex_stats:incr(message_received_malformed),
+                 logplex_realtime:incr('message.received-malformed')
+             end || {malformed, Msg} <- Msgs],
+            RawMsgs = get_raw_msgs([Msg || {Tag, Msg} <- Msgs, Tag =/= malformed]),
+            ChannelMsgsMap = map_msgs_to_channels(RawMsgs),
+            maps:map(fun({ChannelId, Token}, RawChannelMsgs) ->
+                             TokenId = logplex_token:id(Token),
+                             TokenName = logplex_token:name(Token),
+                             process_msgs_batch_redis(lists:reverse(RawChannelMsgs), ChannelId, TokenId, TokenName)
```

Comment:
This `reverse` here is wrong, foldr processes from the list's end but the way we construct the lists inside the map implicitly reverses that.